fluent/protocol/chunk.go

```diff
@@ -0,0 +1,116 @@
+package protocol
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/tinylib/msgp/msgp"
+)
+
+func makeChunkID() (string, error) {
+	b, err := uuid.New().MarshalBinary()
+	if err != nil {
+		return "", err
+	}
+
+	return base64.StdEncoding.EncodeToString(b), nil
+}
+
+//msgp:ignore GzipCompressor
+type ChunkReader struct {
+	br *bytes.Reader
+	R  *msgp.Reader
+}
+
+func (cr *ChunkReader) Reset(b []byte) {
+	if cr.br == nil {
+		cr.br = bytes.NewReader(b)
+		cr.R = msgp.NewReader(cr.br)
+
+		return
+	}
+
+	cr.br.Reset(b)
+	cr.R.Reset(cr.br)
+}
+
+var chunkKeyBits = []byte("chunk")
+
+// GetChunk searches a marshaled message for the "chunk"
+// option value and returns it. The chunk can be used for
+// ack checks without the overhead of unmarshalling.
+// GetChunk returns an error if no value is found.
+func GetChunk(b []byte) (string, error) {
+	chunkReader := chunkReaderPool.Get().(*ChunkReader)
+	chunkReader.Reset(b)
```

Comment:
Same as https://github.com/IBM/fluent-forward-go/pull/32#discussion_r769636019

Reply:
`Reset` needs to be called here because the ChunkReader's `bytes.Reader` needs the `b []byte` values to reset itself. https://github.com/IBM/fluent-forward-go/pull/32/files#diff-e47a42a1c82dbfd53bd77c4e8cfbb418a6a9a880085c82108f86b58d35ca3454R36

I changed the `GzipWriter` to match this pattern.

It's not uncommon in Go for `Reset` to get called before the next use, rather than after the last use. E.g., you can see that the `echo` project follows a similar pattern: https://github.com/labstack/echo/blob/master/echo.go#L645